### PR TITLE
set net.ipv4.tcp_retries2=8 to mitigate broken tcp connections

### PR DESCRIPTION
**Pull request by dmitsh**
Configure TCP on the nodes to reduce number of synack_retries.
This will mitigate "broken" tcp connections, resulting with agent nodes being in "NotReady" state for 15-17 minutes. 
This PR addresses issue #1032

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/1044)

### CODE REVIEW

**Comment by seanknox** [comment]
@dmitsh What were your findings with this enabled?

**Comment by dmitsh** [comment]
@acs-bot test this please

**Comment by dmitsh** [comment]
With this change the node goes into "NotReady" state for 75 seconds, then gets back to "Ready".

**Comment by lachie83** [comment]
Can we document this somewhere? I don't want this getting lost in the ether. Also reference the upstream issue.

**Comment by seanknox** [comment]
Discussed in standup that because this is such a significant change that we'll need to monitor for a bit, we're going to merge this after we cut the v0.4.0 release.

**Comment by seanknox** [comment]
75s is much better than 17m. :) Agree that we need to document this somewhere in `docs/` and then this should be good to go.

**Comment by dmitsh** [comment]
I confirmed that the attached disk hasn't been offloaded within 75 seconds.

**Comment by seanknox** [comment]
What does "offloaded + disk" mean in this case? Lachie and I were wondering.

**Comment by dmitsh** [comment]
I was using Anthony's terminology. By "offloaded + disk" I meant that the pod hasn't been rescheduled to run on another node, and therefore the disk hasn't been detached and reattached to another node.

File: docs/kubernetes.md [hunk 1]
```diff
@@ -197,16 +197,25 @@ read and **write** permissions to the target Subscription.
...
+While [Managed disks](../examples/disks-managed/README.md) are supported for the node OS disks, they are currently not supported for persistent volumes. See https://github.com/kubernetes/kubernetes/pull/46360 for details.
+
+## Known issues and mitigations
+
+### Node "NotReady" due to lost TCP connection
+
+Nodes might appear in the "NotReady" state for approx. 15 minutes if master stops receiving updates from agents.
+This is a known upstream kubernetes [issue #41916](https://github.com/kubernetes/kubernetes/issues/41916#issuecomment-312428731). This fixing PR is currently under review.
+
+To mitigate this problem in ACS-Engine, we implemented the fix in the kubelet service by setting **net.ipv4.tcp_retries2=8** prior to starting the kubelet.
```
Comment (seanknox) [inline docs/kubernetes.md]: Change to: 

> ACS-Engine partially mitigates this issue on Linux by detecting dead TCP connections more quickly via **net.ipv4.tcp_retries2=8**.

File: parts/kuberneteskubelet.service [hunk 2]
```diff
@@ -11,6 +11,7 @@ ExecStartPre=/bin/bash /opt/azure/containers/kubelet.sh
 ExecStartPre=/bin/mkdir -p /var/lib/kubelet
 ExecStartPre=/bin/bash -c "if [ $(mount | grep \"/var/lib/kubelet\" | wc -l) -le 0 ] ; then /bin/mount --bind /var/lib/kubelet /var/lib/kubelet ; fi"
 ExecStartPre=/bin/mount --make-shared /var/lib/kubelet
+ExecStartPre=/sbin/sysctl -w net.ipv4.tcp_retries2=8
```
Comment (seanknox) [inline parts/kuberneteskubelet.service]: Can you add this comment above this line:
```
# This is a partial workaround to this upstream Kubernetes issue: https://github.com/kubernetes/kubernetes/issues/41916#issuecomment-312428731
ExecStartPre=/sbin/sysctl -w net.ipv4.tcp_retries2=8
```

**Comment by lachie83** [comment]
Can you fix the couple of items @seanknox suggested then I think we are good to go

**Comment by dmitsh** [comment]
done

**Comment by lachie83** [comment]
LGTM

**Pull request merged**